### PR TITLE
Fix inconsistent state reporting on the resource endpoints

### PR DESCRIPTION
**Pull request by arnaudsjs**
# Description

Fix bug where the `GET /api/v2/resource/<rid>` and `GET /api/v2/resource` endpoints return an incorrect resourcestate if a resource moved back to the available state in a new version of the configurationmodel.

# Self Check:

- [ ] ~~Attached issue to pull request~~
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] ~~End user documentation is included or an issue is created for end-user documentation~~
- [ ] ~~If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

File: src/inmanta/data/__init__.py [hunk 1]
```diff
@@ -4641,34 +4641,40 @@ async def get_current_resource_state(cls, env: uuid.UUID, rid: ResourceIdStr) ->
...
-                WHERE rps.environment=$1 AND rps.resource_id=$2
-            ) AS last_non_deploying_status
+                CASE
+                    -- The resource_persistent_state.last_non_deploying_status column is only populated for
+                    -- actual deployment operations to prevent locking issues. This case-statement calculates
+                    -- the correct state from the combination of the resource table and the
+                    -- resource_persistent_state table.
+                    WHEN r.status::text = 'deploying'
+                        THEN r.status::text
+                    WHEN r.status = 'available' AND rps.last_non_deploying_status IN('undefined', 'skipped_for_undefined')
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:4658]: nice

**Review by sanderr — COMMENTED** [review]
Commenting so you can rerequest when the dust has settled, i.e. the [discussion on Slack](https://inmanta.slack.com/archives/CKRF0C8R3/p1715161636398029) and associated change(s).

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -4641,34 +4641,39 @@ async def get_current_resource_state(cls, env: uuid.UUID, rid: ResourceIdStr) ->
...
+                    -- actual deployment operations to prevent locking issues. This case-statement calculates
+                    -- the correct state from the combination of the resource table and the
+                    -- resource_persistent_state table.
+                    WHEN r.status::text IN('deploying', 'undefined', 'skipped_for_undefined')
+                        -- The deploying, undefined and skipped_for_undefined states are not tracked in the
+                        -- resource_persistent_state table.
+                        THEN r.status::text
+                    WHEN rps.last_deployed_attribute_hash != r.attribute_hash
+                        -- The hash changed since the last deploy -> new desired state
+                        THEN 'available'
```
Comment (sanderr) [inline src/inmanta/data/__init__.py:4662]: Could this be `unavailable` as well? i.e. do we consider that "deployed" for the sake of the hash update? We may want to simply return `r.status::text` in this case as well, or am I missing something?

**Review by sanderr — COMMENTED** [review]
I'm adding a link to the related [Slack discussion](https://inmanta.slack.com/archives/CKRF0C8R3/p1715161636398029) here because I bet I'll come looking for its context through git blame at some point.

I'll also release comments as I go to speed up the process.

File: src/inmanta/data/dataview.py [hunk 3]
```diff
@@ -562,12 +562,21 @@ def get_base_query(self) -> SimpleQueryBuilder:
...
+                                    -- The deploying, undefined and skipped_for_undefined states are not tracked in the
+                                    -- resource_persistent_state table.
                                     THEN r.status::text
-                                ELSE
-                                    rps.last_non_deploying_status::text
+                                WHEN rps.last_deployed_attribute_hash != r.attribute_hash
+                                   -- The hash changed since the last deploy -> new desired state
+                                   THEN 'available'
+                                   -- No override required, use last known state from actual deployment
+                                   ELSE rps.last_non_deploying_status::text
```
Comment (sanderr) [inline src/inmanta/data/dataview.py:579]: I'm a bit scared of the performance impact of this one. I'm writing down that we have to check on mgmt after updating.

**Review by sanderr — CHANGES_REQUESTED** [review]
Looks mostly good, added a few minor comments.

File: src/inmanta/data/__init__.py (hunk 2 above)
Comment (arnaudsjs) [reply src/inmanta/data/__init__.py:4662]: I think we can return `r.status::text` in this case.

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**